File: 10_adapter_part1.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <sstream>
#include <vector>
using std::cout;
using std::endl;
using std::vector;
using std::string;
using std::ifstream;
using std::sort;

void print_vec(vector<int> vec) {
	for (int i: vec) {
		cout << i << ' ';
	}
	cout << "\n" << endl;
}

int main() {
	ifstream myfile;
	myfile.open("input_files/problem10.txt");
	string mytext;

	vector<int> nums;

	int n = 0;

	while (getline (myfile, mytext)) {
		int num = stoi(mytext);
		nums.push_back(num);
		n++;
	}

	sort(nums.begin(), nums.end());

	print_vec(nums);

	// add charging outlet to front
	nums.insert(nums.begin(), 0);

	print_vec(nums);


	int i = 1;

	int jolt_diff1 = 0;;
	int jolt_diff3 = 0;;

	while (i < nums.size()) {
		int diff = nums[i] - nums[i-1];
		if (diff == 1) {
			jolt_diff1 += 1;
		} else if (diff == 3) {
			jolt_diff3 += 1;
		}
		i++;
	}

	// add 3 jolt diff at end
	jolt_diff3 += 1;

	cout << jolt_diff1 << endl;
	cout << jolt_diff3 << endl;
	cout << "Answer to part 1 " << jolt_diff1 * jolt_diff3 << endl;
}

File: 10_adapter_part1.py
from collections import Counter

file = "input_files/problem10.txt"


data = []
with open(file, "r") as f:
    for line in f:
        l = int(line.replace("\n", ""))
        data.append(l)

print(data)

# first sort
ds = sorted(data)

# add charging outlet to front
ds = [0] + ds

i = 1
diffs = []
while i < len(ds):
    diffs.append(ds[i] - ds[i-1]) 
    i += 1

# add last 3 volt diff
diffs.append(3)

assert any([x>3 for x in diffs]) == False

c = Counter(diffs)
print(c)
answer = c[1] * c[3]
print(f"the answer to part 1 is {answer}")